File: scripts/test-ipos-session-based.py
#!/usr/bin/env python3
"""
iPOS Session-Based API Access Test
This script proves that we CAN access the API by maintaining the browser session.

Key Insight: Tokens are session-bound!
- Login in browser → tokens are valid in that session
- Extract tokens → use elsewhere → tokens expire/invalid

Solution: Keep the browser session alive and make API calls from within it.
"""

import asyncio
import json
from playwright.async_api import async_playwright
from datetime import datetime

# Configuration (from .env or hardcoded for testing)
IPOS_EMAIL = "sipnplay@ipos.vn"
IPOS_PASSWORD = "123123A"
IPOS_BRAND_UID = "32774afe-fd5c-4028-b837-f91837c0307c"
IPOS_COMPANY_UID = "8a508e04-440f-4145-9429-22b7696c6193"
IPOS_STORE_UID = "72a800a6-1719-4b4b-9065-31ab2e0c07e5"
STORE_OPEN_HOUR = 10


def get_store_opening_timestamp(date=None):
    """Get timestamp in milliseconds for store opening time (10 AM)"""
    if date is None:
        date = datetime.now()

    opening = date.replace(hour=STORE_OPEN_HOUR, minute=0, second=0, microsecond=0)
    return int(opening.timestamp() * 1000)


async def test_session_based_api_access():
    """
    Test accessing iPOS API by maintaining browser session.
    This proves the concept before implementing in Node.js/TypeScript.
    """

    print("\n" + "=" * 60)
    print("iPOS Session-Based API Access Test")
    print("=" * 60 + "\n")

    captured_api_data = None

    async with async_playwright() as p:
        # Launch browser (headless for production, headed for debugging)
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context()
        page = await context.new_page()

        # Enable network logging
        def handle_response(response):
            nonlocal captured_api_data

            # Look for the sale-summary/overview API call
            if "sale-summary/overview" in response.url:
                print(f"\n🎯 CAPTURED API CALL: {response.url}")
                print(f"   Status: {response.status}")

                # Try to get the JSON response
                try:
                    if response.status == 200:
                        # We need to capture this asynchronously
                        async def capture():
                            nonlocal captured_api_data
                            try:
                                data = await response.json()
                                captured_api_data = data
                                print(f"   ✅ Successfully captured API response!")
                                print(f"   Data preview: {json.dumps(data, indent=2)[:200]}...")
                            except Exception as e:
                                print(f"   ⚠️  Could not parse JSON: {e}")

                        # Schedule the capture
                        asyncio.create_task(capture())
                except Exception as e:
                    print(f"   Error: {e}")

        page.on("response", handle_response)

        print("Step 1: Navigating to login page...")
        await page.goto("https://fabi.ipos.vn/login")

        print("Step 2: Filling in credentials...")
        await page.fill('input[name="email_input"]', IPOS_EMAIL)
        await page.fill('input[type="password"]', IPOS_PASSWORD)

        print("Step 3: Logging in...")
        # Click login and wait for navigation
        async with page.expect_navigation():
            await page.click('button:has-text("Đăng nhập")')

        print("Step 4: Waiting for dashboard to load...")
        try:
            await page.wait_for_selector('text=Doanh thu (NET)', timeout=15000)
            print("   ✓ Dashboard loaded successfully")
        except:
            print("   ⚠ Dashboard selector not found, continuing anyway...")

        print("Step 5: Waiting for API calls to complete...")
        await page.wait_for_timeout(5000)

        print("\n" + "=" * 60)
        print("TEST RESULTS")
        print("=" * 60)

        if captured_api_data:
            print("\n✅ SUCCESS! API data captured from browser session!\n")

            # Extract the important fields
            data = captured_api_data.get("data", {})
            sale_tracking = data.get("sale_tracking", {})

            print("📊 Dashboard Data:")
            print(f"   Unpaid Amount: {sale_tracking.get('total_amount', 0):,} VND")
            print(f"   Today's Revenue (NET): {data.get('revenue_net', 0):,} VND")
            print(f"   Active Tables: {sale_tracking.get('table_count', 0)}")
            print(f"   Current Customers: {sale_tracking.get('people_count', 0)}")

            print("\n🎉 PROOF OF CONCEPT SUCCESSFUL!")
            print("\n💡 Key Findings:")
            print("   1. We CAN access the API by maintaining browser session")
            print("   2. No need to manually capture and manage tokens")
            print("   3. This approach will work on Railway in production")

            print("\n📋 Next Steps:")
            print("   1. Convert this logic to Node.js/TypeScript")
            print("   2. Create a service that maintains browser session")
            print("   3. Extract data from API responses (not HTML)")
            print("   4. Deploy to Railway staging for testing")

        else:
            print("\n❌ FAILED: No API data captured")
            print("\nPossible reasons:")
            print("   1. Dashboard didn't load the API call")
            print("   2. API call was made but response wasn't captured")
            print("   3. Network timing issue")

            print("\n💡 Debugging tips:")
            print("   1. Run with headless=False to see what's happening")
            print("   2. Check network tab in browser DevTools")
            print("   3. Increase wait time after login")

        print("\n" + "=" * 60 + "\n")

        # Keep browser open for manual inspection
        if not captured_api_data:
            print("Browser will stay open for 30 seconds for debugging...")
            await page.wait_for_timeout(30000)

        await browser.close()


if __name__ == "__main__":
    asyncio.run(test_session_based_api_access())

File: scripts/test-ipos-session-based.js
// Test script for session-based iPOS API access
// This proves the concept works before deploying to production

const { chromium } = require('playwright');
require('dotenv').config();

const IPOS_EMAIL = process.env.IPOS_EMAIL || 'sipnplay@ipos.vn';
const IPOS_PASSWORD = process.env.IPOS_PASSWORD;

if (!IPOS_PASSWORD) {
  console.error('ERROR: IPOS_PASSWORD not set in .env file');
  process.exit(1);
}

async function testSessionBasedAPI() {
  console.log('\n' + '='.repeat(60));
  console.log('iPOS Session-Based API Access Test');
  console.log('='.repeat(60) + '\n');

  let capturedData = null;

  const browser = await chromium.launch({ headless: false });
  const context = await browser.newContext();
  const page = await context.newPage();

  // Set up response listener BEFORE navigating
  page.on('response', async (response) => {
    if (response.url().includes('sale-summary/overview')) {
      console.log(`\n🎯 CAPTURED API CALL: ${response.url()}`);
      console.log(`   Status: ${response.status()}`);

      if (response.status() === 200) {
        try {
          const data = await response.json();
          capturedData = data;
          console.log('   ✅ Successfully captured API response!');
          console.log(`   Data preview: ${JSON.stringify(data, null, 2).substring(0, 200)}...`);
        } catch (error) {
          console.error('   ⚠️  Error parsing JSON:', error.message);
        }
      } else {
        const text = await response.text();
        console.error(`   ❌ Non-200 response: ${text.substring(0, 200)}`);
      }
    }
  });

  try {
    console.log('Step 1: Navigating to login page...');
    await page.goto('https://fabi.ipos.vn/login', { waitUntil: 'networkidle' });

    console.log('Step 2: Filling in credentials...');
    await page.fill('input[name="email_input"]', IPOS_EMAIL);
    await page.fill('input[type="password"]', IPOS_PASSWORD);

    console.log('Step 3: Logging in...');
    await Promise.all([
      page.waitForNavigation({ waitUntil: 'networkidle' }),
      page.click('button:has-text("Đăng nhập")')
    ]);

    console.log('Step 4: Waiting for dashboard to load...');
    try {
      await page.waitForSelector('text=Doanh thu (NET)', { timeout: 15000 });
      console.log('   ✓ Dashboard loaded successfully');
    } catch (e) {
      console.log('   ⚠ Dashboard selector not found, continuing anyway...');
    }

    console.log('Step 5: Waiting for API calls to complete...');
    console.log('   Waiting 5 seconds for API calls...');
    await page.waitForTimeout(5000);

    console.log('\n' + '='.repeat(60));
    console.log('TEST RESULTS');
    console.log('='.repeat(60));

    if (capturedData) {
      console.log('\n✅ SUCCESS! API data captured from browser session!\n');

      // Extract the important fields
      const data = capturedData.data || {};
      const saleTracking = data.sale_tracking || {};

      console.log('📊 Dashboard Data:');
      console.log(`   Unpaid Amount: ${formatVND(saleTracking.total_amount || 0)} VND`);
      console.log(`   Today's Revenue (NET): ${formatVND(data.revenue_net || 0)} VND`);
      console.log(`   Active Tables: ${saleTracking.table_count || 0}`);
      console.log(`   Current Customers: ${saleTracking.people_count || 0}`);

      console.log('\n🎉 PROOF OF CONCEPT SUCCESSFUL!');
      console.log('\n💡 Key Findings:');
      console.log('   1. We CAN access the API by maintaining browser session');
      console.log('   2. No need to manually capture and manage tokens');
      console.log('   3. This approach will work on Railway in production');

      console.log('\n📋 Next Steps:');
      console.log('   1. Use ipos-session-service.ts in production');
      console.log('   2. Update /api/pos/dashboard to use session service');
      console.log('   3. Deploy to Railway staging for testing');
      console.log('   4. Monitor for session stability');

      // Test immediate re-use of session
      console.log('\n' + '='.repeat(60));
      console.log('TESTING IMMEDIATE SESSION RE-USE');
      console.log('='.repeat(60));

      console.log('\nNow testing if we can make another API call in the same session...');
      console.log('Reloading page to trigger another API call...');

      let secondCapturedData = null;
      const secondDataPromise = new Promise((resolve) => {
        const handler = async (response) => {
          if (response.url().includes('sale-summary/overview')) {
            if (response.status() === 200) {
              try {
                const data = await response.json();
                secondCapturedData = data;
                page.off('response', handler);
                resolve(data);
              } catch (e) {
                console.error('Error parsing second response:', e.message);
              }
            }
          }
        };
        page.on('response', handler);
      });

      await page.reload({ waitUntil: 'networkidle' });
      await Promise.race([
        secondDataPromise,
        new Promise((resolve) => setTimeout(resolve, 10000))
      ]);

      if (secondCapturedData) {
        console.log('\n✅ SUCCESS! Second API call also works in same session!');
        console.log('   This proves session persistence works.\n');
      } else {
        console.log('\n⚠️  Second API call not captured (may need longer wait)\n');
      }

    } else {
      console.log('\n❌ FAILED: No API data captured');
      console.log('\nPossible reasons:');
      console.log('   1. Dashboard didn\'t load the API call');
      console.log('   2. API call was made but response wasn\'t captured');
      console.log('   3. Network timing issue');

      console.log('\n💡 Debugging tips:');
      console.log('   1. Browser is staying open - check network tab');
      console.log('   2. Look for sale-summary/overview requests');
      console.log('   3. Check if any errors in console');
    }

    console.log('\n' + '='.repeat(60) + '\n');

    // Keep browser open for manual inspection
    if (!capturedData) {
      console.log('Browser will stay open for 30 seconds for debugging...');
      await page.waitForTimeout(30000);
    } else {
      console.log('Keeping browser open for 10 seconds...');
      await page.waitForTimeout(10000);
    }

  } catch (error) {
    console.error('\n❌ ERROR:', error);
  } finally {
    await browser.close();
  }
}

function formatVND(amount) {
  return new Intl.NumberFormat('vi-VN').format(amount);
}

testSessionBasedAPI().catch(console.error);
